storage/innobase/btr/btr0cur.cc

```diff
@@ -971,68 +970,77 @@ static int btr_latch_prev(buf_block_t *block, page_id_t page_id,
   buffer-fixes on both blocks will prevent eviction. */
 
  retry:
-  /* Pass no_wait pointer to ensure that we don't wait on the current page
-  latch while holding the next page latch to avoid latch ordering violation. */
-  bool no_wait= false;
   int ret= 1;
-
-  buf_block_t *prev= buf_page_get_gen(page_id, zip_size, RW_NO_LATCH, nullptr,
-                                      BUF_GET, mtr, err, false, &no_wait);
+  buf_block_t *prev= buf_pool.page_fix(page_id, err, true);
   if (UNIV_UNLIKELY(!prev))
-  {
-    /* Check if we had to return because we couldn't wait on latch. */
-    if (no_wait)
-      goto ordered_latch;
     return 0;
-  }
+  if (prev == reinterpret_cast<buf_block_t*>(-1))
```

Comment:
1. Most callers are not interested in an error code. Only `btr_latch_prev()` is (or actually its grandcallers are). In any ISA that I am aware of, the return value is guaranteed to be in a register. If the primary output (the block descriptor) were returned via a pointer, then each caller would not only have to spill that variable on the stack (for the call) but also load the value back from stack. Since you pointed this out, I noticed that we’d have an even less call overhead if we reordered the parameters of `btr_latch_prev()` so that they’d more closely match the parameters of `buf_pool_t::page_fix()`. After I shuffled of the parameters, the second System V AMD64 ABI parameter `rsi` (right after `rdi=this=&buf_pool`) would have be at the right place. I’d invite you to check how much more code your suggestion would generate.
1. Only `btr_latch_prev()` is interested in error codes. Most callers do not pass `no_wait=true` and therefore do not need to deal with the return value `-1`.
1. I am not a friend of boilerplate code. I think that it is better to write `true` than `buf_pool_t::WAIT`. I can write an inline wrapper function that will pass the default parameters for most callers.

Reply:
I now see that `PageConverter::operator()` needs to pass a special mode where also `FREED` blocks will be returned. The `bool no_wait` would thus have to be replaced with a multi-valued `enum`.

Reply:
7f1fbc69fc54c2f175027493829e05900e254063 addresses the third point.